### PR TITLE
feat: passing parameters to actions via the URL

### PR DESCRIPTION
**Pull request by NetaMolcho**

### CODE REVIEW

File: src/actions/controller.ts [hunk 1]
```diff
@@ -29,16 +29,38 @@ export class ActionsController {
     return ActionsController.actions.find(action => action.id === id);
   }
 
-  public static async performActionsFromParams() {
+  public static async performActionsFromURL() {
     const actionsParam = await getParameter("actions");
-    const actionsIds = uniq(compact(split(actionsParam, ",")));
-    getLogger().trace(`configuration - actionsIds= ${actionsIds}`, { method: "performActionsFromParams" });
+    getLogger().trace(`actionsParam= ${actionsParam}`, { method: "performActionsFromURL" });
+    if (actionsParam?.startsWith("%5B") || actionsParam?.startsWith("[")) {
```
Comment (bd82) [inline src/actions/controller.ts:35]: can you add a one line comment how the action parameter looks in each branch, e.g:
`// actions=[{},{}]`
`// actions=foo,bar`

File: src/actions/controller.ts [hunk 2]
```diff
@@ -29,16 +29,38 @@ export class ActionsController {
...
+      ActionsController.perfomFullActions(actionsParam);
```
Comment (bd82) [inline src/actions/controller.ts:36]: I will try to think of a new name to emphasize this action is:
- defined on the fly.
- fully specified.
-  does not match an existing action

Comment (bd82) [reply src/actions/controller.ts:36]: maybe in-lined?

File: tests/controller.spec.ts [hunk 3]
```diff
@@ -166,4 +166,44 @@ describe("controller unit test", () => {
             }
         });
     });
+
+    describe('perfomFullActions', () => {
+        let requireMock: any;
+        before(() => {
```
Comment (bd82) [inline tests/controller.spec.ts:172]: should we use `before**each**` and `after**each**` to increase test case isolation?

File: src/actions/controller.ts (hunk 2 above)
Comment (NetaMolcho) [reply src/actions/controller.ts:36]: renamed to perfomInlinedActions

File: tests/controller.spec.ts (hunk 3 above)
Comment (NetaMolcho) [reply tests/controller.spec.ts:172]: This only matters for the first unit test

File: src/actions/controller.ts [hunk 4]
```diff
@@ -29,16 +28,63 @@ export class ActionsController {
...
-    getLogger().trace(`configuration - actionsIds= ${actionsIds}`, { method: "performActionsFromParams" });
+    if (actionsParam === undefined) {
+      return;
+    }
+    const decodedActionsParam = decodeURI(actionsParam);
+    getLogger().trace(`decodedActionsParam= ${decodedActionsParam}`, { method: "performActionsFromURL" });
+    const mode = ActionsController.detectActionMode(decodedActionsParam);
+    switch (mode) {
+      case "IDs": {
+        const actionsIds = uniq(compact(split(decodedActionsParam, ",")));
```
Comment (bd82) [inline src/actions/controller.ts:41]: we need to trim the array of IDs

File: src/actions/controller.ts [hunk 5]
```diff
@@ -29,16 +28,63 @@ export class ActionsController {
...
+        ActionsController.performActionsIds(actionsIds);
+        break;
+      }
+      case "Inlined": {
+        ActionsController.perfomInlinedActions(decodedActionsParam.trim());
```
Comment (bd82) [inline src/actions/controller.ts:46]: - we don't need to trim here, JSON.parse() does it.
- we should pass the decoded string  here

File: src/actions/controller.ts (hunk 4 above)
Comment (bd82) [reply src/actions/controller.ts:41]: I saw you did it inside: `performActionsIds`
but then we have the "processing" of input split into two places.
I'd rather do it in one place only

File: src/actions/controller.ts [hunk 6]
```diff
@@ -29,16 +28,63 @@ export class ActionsController {
...
+      //actions=openSettings,openGoogle
+    }
+    return "IDs";
+  }
+
+  private static performActionsIds(actionsIds: string[]) {
+    getLogger().trace(`actionsIds= ${actionsIds}`, { method: "performActionsIds" });
       forEach(actionsIds, async actionId => {
-      const action = ActionsController.getAction(actionId);
+      const action = ActionsController.getAction(actionId.trim());
```
Comment (bd82) [inline src/actions/controller.ts:70]: do the trimming with the other processing (uniq/concat/...) so the "processing" is handled in a **single** place.

File: src/actions/controller.ts [hunk 7]
```diff
@@ -29,16 +28,63 @@ export class ActionsController {
...
+      }
+    }
+    catch (e) {
+      // actionsIDsMode
+      //actions=openSettings,openGoogle
+    }
+    return "IDs";
+  }
+
+  private static performActionsIds(actionsIds: string[]) {
```
Comment (bd82) [inline src/actions/controller.ts:67]: what do you think about naming this function: performActions**By**Ids

So we have:
- perform**Inlined**Actions
- performActions**ByIds**

Which makes more sense from English grammar perspective.

File: src/actions/controller.ts (hunk 4 above)
Comment (bd82) [reply src/actions/controller.ts:41]: subjective because we immediately use it.

File: tests/controller.spec.ts (hunk 3 above)
Comment (bd82) [reply tests/controller.spec.ts:172]: We still need good isolation between tests to avoid none deterministic behavior.
For example if we do `.only` for the second test these `before` and `after` blocks would still run
can we guarantee they would not affect the results?

File: src/actions/controller.ts (hunk 6 above)
Comment (bd82) [reply src/actions/controller.ts:70]: decided this is subjective 👍

**Review by bd82 — APPROVED** [review]
